### PR TITLE
added Mixins for last viewed products

### PR DESCRIPTION
**Pull request by volodin28**

### CODE REVIEW

File: src/apps/inventory/mixins.py [hunk 1]
```diff
@@ -22,3 +22,54 @@ def get_context_data(self, **kwargs):
...
+    def get_last_viewed_products(self):
+        redis = get_redis_connection("redis_cache")
+        user = self.request.user
+        if user.is_authenticated:
+            user_key = f"user:{user.uuid}:recentviews"
+        else:
+            user_key = f"session:{self.request.session.session_key}:recentviews"
+
+        last_viewed_products_ids = redis.zrevrange(user_key, 0, 2)
+        products = Product.objects.in_bulk(map(int, last_viewed_products_ids))
```
Comment (egorvavilov) [inline src/apps/inventory/mixins.py]: Nice

File: src/apps/inventory/mixins.py [hunk 2]
```diff
@@ -22,3 +22,54 @@ def get_context_data(self, **kwargs):
...
+        ordered_products = OrderedDict(
+            (int(product_id), products[int(product_id)]) for product_id in last_viewed_products_ids
```
Comment (egorvavilov) [inline src/apps/inventory/mixins.py]: Why so complicated? You already extracted products and mapped them to ids with in_bulk query

File: src/apps/inventory/mixins.py [hunk 3]
```diff
@@ -22,3 +22,54 @@ def get_context_data(self, **kwargs):
...
+        user = self.request.user
+
+        if user.is_authenticated:
+            user_key = f"user:{user.uuid}:recentviews"
+        else:
+            user_key = f"session:{request.session.session_key}:recentviews"
+
+        redis = get_redis_connection("redis_cache")
+
+        view_count = redis.incr(f"{user_key}:counter")
```
Comment (egorvavilov) [inline src/apps/inventory/mixins.py]: Can we simply dump datetime in seconds smth like "123214234525" ?
Set will guarantee uniquness and timecode will ensure lastviewed logic

File: src/apps/inventory/mixins.py [hunk 4]
```diff
@@ -22,3 +22,54 @@ def get_context_data(self, **kwargs):
...
+        redis.zadd(user_key, {product.id: view_count})
```
Comment (egorvavilov) [inline src/apps/inventory/mixins.py]: you don't need to add it
because you already incr your value

File: src/apps/inventory/mixins.py (hunk 3 above)
Comment (egorvavilov) [reply src/apps/inventory/mixins.py]: wait, now I don't get it at all

**Review by egorvavilov — CHANGES_REQUESTED** [review]

**Comment by GrazhdanochkaMariya** [comment]
We started adding some documentation to the project and your complex task definitely needs in it

**Review by egorvavilov — APPROVED** [review]

**Pull request merged**